core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -3523,15 +3523,37 @@ object Stream extends StreamLowPriority {
     * All elements that are available, up to the specified limit,
     * are dequeued and emitted as a single chunk.
     */
-  def fromQueueUnterminated[F[_]: Functor, A](
+  def fromQueueUnterminated[F[_], A](
       queue: QueueSource[F, A],
       limit: Int = Int.MaxValue
-  ): Stream[F, A] =
-    fromQueueNoneTerminatedSingletons_[F, A](
-      queue.take.map(a => Some(a)),
-      queue.tryTake.map(_.map(a => Some(a))),
-      limit
-    )
+  )(implicit F: Functor[F]): Stream[F, A] =
+    F match {
+      case f0: Async[F] =>
```

Comment:
We don't need `Async` here. In fact, I propose we replace this entire implementation here with:

```scala
Stream.eval(queue.tryTakeN(someLimit)).flatMap {
  case Nil => Stream.eval(queue.take)
  case as => Stream.emits(as)
}.repeat
```

Reply:
I believe this will drain the queue indefinitely, as `someLimit` will not reflect the already pulled elements.

Reply:
How is the existing logic different?
https://github.com/typelevel/fs2/blob/8cf9273bf9f99f32f7989724ca4564a6b8f3ac03/core/shared/src/main/scala/fs2/Stream.scala#L3540-L3548

Reply:
> We don't need `Async` here. In fact, I propose we replace this entire implementation here with:

Not really, `tryTakeN` expects at least a `Monad`:
```scala
def tryTakeN(maxN: Option[Int])(implicit F: Monad[F])
```

Reply:
Ok, so we need `Monad`, not `Async` :)

Reply:
> How is the existing logic different?

My bad, same same. 

The rationale behind `Async` was that only a bounded async queue sports an optimized `tryTakeN`. I like your proposition, (as it simplifies the code substantially), just to make sure I might benchmark it against current solution, which uses `acc: collection.mutable.Builder[A, collection.mutable.Buffer[A]]`, as opposed to a `List`.

Reply:
> The rationale behind `Async` was that only a bounded async queue sports an optimized `tryTakeN`

True, but checking if `F` is `Async` actually doesn't tell us if the `Queue` is async or not, we don't really have a good way to check that.

I think it is better to use the most optimal method in the `Queue` API. The onus should be on the `Queue` implementation to be fast.

Reply:
I updated the code. 

If the signature 
```scala
  def fromQueueUnterminated[F[_], A](
      queue: QueueSource[F, A],
      limit: Int = Int.MaxValue
  )(implicit F: Functor[F])
```
must not change to 
```scala
  def fromQueueUnterminated[F[_], A](
      queue: QueueSource[F, A],
      limit: Int = Int.MaxValue
  )(implicit F: Monad[F])
```
(source/bin compatibility?), the old implementation has to stay and we need to pattern-match on `F`?